File: apps/minecraft/spigot/src/main/java/crabcraft/net/crabUtilities/slime/SlimeMap.java
package crabcraft.net.crabUtilities.slime;

import crabcraft.net.crabUtilities.CrabMessages;
import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.format.TextDecoration;
import org.bukkit.Bukkit;
import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.World;
import org.bukkit.entity.Player;
import org.bukkit.inventory.Inventory;
import org.bukkit.inventory.InventoryHolder;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.ItemMeta;

import java.util.List;

/** A six-row slime-chunk map oriented in the direction the player is facing. */
final class SlimeMap implements InventoryHolder {

    private static final int WIDTH = 9;
    private static final int SIZE = 54;
    private static final int CENTER_COLUMN = 4;
    private static final int CENTER_ROW = 3;
    private static final int[] X_ROTATION = { 1, 0, -1, 0 };
    private static final int[] Z_ROTATION = { 0, 1, 0, -1 };

    private final Inventory inventory;

    private SlimeMap(Player player) {
        this.inventory = Bukkit.createInventory(
                this, SIZE, CrabMessages.accent("Slime Chunks"));
        populate(player);
    }

    static void open(Player player) {
        SlimeMap map = new SlimeMap(player);
        player.openInventory(map.inventory);
    }

    private void populate(Player player) {
        Location location = player.getLocation();
        World world = player.getWorld();
        int centerChunkX = location.getBlockX() >> 4;
        int centerChunkZ = location.getBlockZ() >> 4;

        for (int slot = 0; slot < SIZE; slot++) {
            ChunkOffset offset = chunkOffsetAt(location.getYaw(), slot);
            int chunkX = centerChunkX + offset.x();
            int chunkZ = centerChunkZ + offset.z();
            boolean currentChunk = offset.x() == 0 && offset.z() == 0;
            boolean slimeChunk = world.getChunkAt(chunkX, chunkZ, false).isSlimeChunk();
            inventory.setItem(slot, createMapItem(chunkX, chunkZ, currentChunk, slimeChunk));
        }
    }

    private static ItemStack createMapItem(int chunkX, int chunkZ, boolean currentChunk, boolean slimeChunk) {
        Material material = currentChunk
                ? Material.BLUE_STAINED_GLASS_PANE
                : slimeChunk ? Material.LIME_STAINED_GLASS_PANE : Material.RED_STAINED_GLASS_PANE;
        ItemStack item = new ItemStack(material);
        ItemMeta meta = item.getItemMeta();

        int blockX = chunkX << 4;
        int blockZ = chunkZ << 4;
        String position = "(" + blockX + ", " + blockZ + ")";
        meta.displayName(plain(currentChunk
                ? CrabMessages.highlight("You are here ")
                        .append(CrabMessages.text(position))
                : CrabMessages.text(position)));
        meta.lore(List.of(
                plain(slimeChunk
                        ? CrabMessages.success("Slime chunk")
                        : CrabMessages.error("Not a slime chunk")),
                plain(CrabMessages.muted("Chunk " + chunkX + ", " + chunkZ))));
        item.setItemMeta(meta);
        return item;
    }

    private static Component plain(Component component) {
        return component.decoration(TextDecoration.ITALIC, false);
    }

    /** Converts one GUI slot into a world chunk offset using the nearest cardinal yaw. */
    static ChunkOffset chunkOffsetAt(float yaw, int slot) {
        if (slot < 0 || slot >= SIZE) {
            throw new IllegalArgumentException("slot must be between 0 and " + (SIZE - 1));
        }

        int screenX = slot % WIDTH - CENTER_COLUMN;
        int screenZ = slot / WIDTH - CENTER_ROW;
        int direction = Math.floorMod(Math.round(yaw / 90.0F) + 2, X_ROTATION.length);
        int xRotation = X_ROTATION[direction];
        int zRotation = Z_ROTATION[direction];

        int worldX = screenX * xRotation - screenZ * zRotation;
        int worldZ = screenX * zRotation + screenZ * xRotation;
        return new ChunkOffset(worldX, worldZ);
    }

    @Override
    public Inventory getInventory() {
        return inventory;
    }

    record ChunkOffset(int x, int z) {}
}

File: apps/minecraft/spigot/src/main/kotlin/crabcraft/net/crabUtilities/slime/SlimeMap.kt
package crabcraft.net.crabUtilities.slime

import crabcraft.net.crabUtilities.CrabMessages
import net.kyori.adventure.text.Component
import net.kyori.adventure.text.format.TextDecoration
import org.bukkit.Bukkit
import org.bukkit.Material
import org.bukkit.entity.Player
import org.bukkit.inventory.Inventory
import org.bukkit.inventory.InventoryHolder
import org.bukkit.inventory.ItemStack

/** A six-row slime-chunk map oriented in the direction the player is facing. */
class SlimeMap private constructor(player: Player) : InventoryHolder {
    private val inventory = Bukkit.createInventory(this, SIZE, CrabMessages.accent("Slime Chunks"))

    init {
        val location = player.location
        val world = player.world
        val centreChunkX = location.blockX shr 4
        val centreChunkZ = location.blockZ shr 4
        for (slot in 0 until SIZE) {
            val offset = chunkOffsetAt(location.yaw, slot)
            val chunkX = centreChunkX + offset.x()
            val chunkZ = centreChunkZ + offset.z()
            val currentChunk = offset.x() == 0 && offset.z() == 0
            val slimeChunk = world.getChunkAt(chunkX, chunkZ, false).isSlimeChunk
            inventory.setItem(slot, createMapItem(chunkX, chunkZ, currentChunk, slimeChunk))
        }
    }

    override fun getInventory(): Inventory = inventory

    data class ChunkOffset(private val x: Int, private val z: Int) {
        fun x() = x

        fun z() = z
    }

    companion object {
        private const val WIDTH = 9
        private const val SIZE = 54
        private const val CENTER_COLUMN = 4
        private const val CENTER_ROW = 3
        private val X_ROTATION = intArrayOf(1, 0, -1, 0)
        private val Z_ROTATION = intArrayOf(0, 1, 0, -1)

        @JvmStatic
        fun open(player: Player) {
            player.openInventory(SlimeMap(player).inventory)
        }

        private fun createMapItem(chunkX: Int, chunkZ: Int, currentChunk: Boolean, slimeChunk: Boolean): ItemStack {
            val material =
                when {
                    currentChunk -> Material.BLUE_STAINED_GLASS_PANE
                    slimeChunk -> Material.LIME_STAINED_GLASS_PANE
                    else -> Material.RED_STAINED_GLASS_PANE
                }
            val item = ItemStack(material)
            val meta = item.itemMeta
            val position = "(${chunkX shl 4}, ${chunkZ shl 4})"
            meta.displayName(
                plain(
                    if (currentChunk) CrabMessages.highlight("You are here ").append(CrabMessages.text(position))
                    else CrabMessages.text(position)
                )
            )
            meta.lore(
                listOf(
                    plain(
                        if (slimeChunk) CrabMessages.success("Slime chunk") else CrabMessages.error("Not a slime chunk")
                    ),
                    plain(CrabMessages.muted("Chunk $chunkX, $chunkZ")),
                )
            )
            item.itemMeta = meta
            return item
        }

        private fun plain(component: Component) = component.decoration(TextDecoration.ITALIC, false)

        @JvmStatic
        fun chunkOffsetAt(yaw: Float, slot: Int): ChunkOffset {
            require(slot in 0 until SIZE) { "slot must be between 0 and ${SIZE - 1}" }
            val screenX = slot % WIDTH - CENTER_COLUMN
            val screenZ = slot / WIDTH - CENTER_ROW
            val direction = Math.floorMod(Math.round(yaw / 90.0f) + 2, X_ROTATION.size)
            val xRotation = X_ROTATION[direction]
            val zRotation = Z_ROTATION[direction]
            return ChunkOffset(screenX * xRotation - screenZ * zRotation, screenX * zRotation + screenZ * xRotation)
        }
    }
}
